csharp/ql/src/semmle/code/csharp/controlflow/internal/Completion.qll

```diff
@@ -668,165 +658,142 @@ abstract private class InheritableCompletion extends AbnormalCompletion { }
-  override string toString() { result = "break" }
-}
+class BreakCompletion extends Completion {
+  BreakCompletion() {
+    this = TBreakCompletion() or
+    this = TNestedCompletion(_, TBreakCompletion())
+  }
 
-private class BreakCompletionInherited extends BreakCompletion, InheritedCompletion {
-  BreakCompletionInherited() { this = TInheritedCompletion(_, TBreakCompletion()) }
+  override string toString() { this = TBreakCompletion() and result = "break" }
 }
 
 /**
  * A completion that represents evaluation of a statement or an
  * expression resulting in a loop continuation (a `continue`
  * statement).
  */
-abstract class ContinueCompletion extends AbnormalCompletion { }
-
-private class ContinueCompletionDirect extends ContinueCompletion, InheritableCompletion,
-  TContinueCompletion {
-  override string toString() { result = "continue" }
-}
+class ContinueCompletion extends Completion {
+  ContinueCompletion() {
+    this = TContinueCompletion() or
+    this = TNestedCompletion(_, TContinueCompletion())
+  }
 
-private class ContinueCompletionInherited extends ContinueCompletion, InheritedCompletion {
-  ContinueCompletionInherited() { this = TInheritedCompletion(_, TContinueCompletion()) }
+  override string toString() { this = TContinueCompletion() and result = "continue" }
 }
 
 /**
  * A completion that represents evaluation of a statement or an
  * expression resulting in a `goto` jump.
  */
-abstract class GotoCompletion extends AbnormalCompletion { }
+abstract class GotoCompletion extends Completion { }
 
 /**
  * A completion that represents evaluation of a statement or an
  * expression resulting in a `goto label` jump.
  */
-abstract class GotoLabelCompletion extends GotoCompletion {
+class GotoLabelCompletion extends GotoCompletion {
+  private GotoLabelStmt goto;
+
+  GotoLabelCompletion() {
+    this = TGotoLabelCompletion(goto) or
+    this = TNestedCompletion(_, TGotoLabelCompletion(goto))
+  }
+
   /** Gets the target of the `goto label` completion. */
   string getLabel() { result = this.getGotoStmt().getLabel() }
 
   /** Gets the statement that resulted in this `goto label` completion. */
-  abstract GotoLabelStmt getGotoStmt();
-}
-
-private class GotoLabelCompletionDirect extends GotoLabelCompletion, InheritableCompletion,
-  TGotoLabelCompletion {
-  override GotoLabelStmt getGotoStmt() { this = TGotoLabelCompletion(result) }
-
-  override string toString() { result = "goto(" + getLabel() + ")" }
-}
-
-private class GotoLabelCompletionInherited extends GotoLabelCompletion, InheritedCompletion {
-  private TGotoLabelCompletion inherited;
+  GotoLabelStmt getGotoStmt() { result = goto }
 
-  GotoLabelCompletionInherited() { this = TInheritedCompletion(_, inherited) }
-
-  override GotoLabelStmt getGotoStmt() { inherited = TGotoLabelCompletion(result) }
+  override string toString() {
+    this = TGotoLabelCompletion(goto) and result = "goto(" + this.getLabel() + ")"
```

Comment:
This is a little sketchy. I suppose, morally, there should be a `or result = this.(NestedCompletion).toString()` to ensure that `toString()` always exists.

Reply:
Ditto for the cases below.

Reply:
I know, how about a comment instead?